exonum/src/api/ext.rs

```diff
@@ -0,0 +1,542 @@
+//! # use exonum::crypto::{Hash, PublicKey};
+//! # use exonum::encoding;
+//! # use exonum::messages::RawTransaction;
+//! # use exonum::storage::{Fork, Snapshot};
+//!
+//! // Transactions
+//! transactions! {
+//!     Any {
+//!         const SERVICE_ID = 1000;
+//!
+//!         struct CreateWallet {
+//!             // ...
+//!         }
+//!
+//!         struct Transfer {
+//!             // ...
+//!         }
+//!     }
+//! }
+//!
+//! impl Transaction for CreateWallet {
+//!     // ...
+//! #   fn verify(&self) -> bool { true }
+//! #   fn execute(&self, _: &mut Fork) -> ExecutionResult { Ok(()) }
+//! }
+//!
+//! impl Transaction for Transfer {
+//!     // ...
+//! #   fn verify(&self) -> bool { true }
+//! #   fn execute(&self, _: &mut Fork) -> ExecutionResult { Ok(()) }
+//! }
+//!
+//! // Service schema containing wallet balances
+//! struct Schema { /* ... */ }
+//!
+//! impl Schema {
+//! #   fn new<S: AsRef<Snapshot>>(snapshot: S) -> Self { Schema { } }
+//!     pub fn balance(&self, key: &PublicKey) -> Option<u64> {
+//!         // ...
+//! #       Some(42)
+//!     }
+//! }
+//!
+//! // Read request: get the balance of a specific wallet.
+//! pub const BALANCE_SPEC: Spec = Spec { id: "balance" };
+//!
+//! pub fn balance(ctx: &ReadContext, key: PublicKey) -> ApiResult<Option<u64>> {
+//!     let schema = Schema::new(ctx.snapshot());
+//!     Ok(schema.balance(&key))
+//! }
+//!
+//! // In `Service` implementation:
+//! # struct MyService;
+//! impl Service for MyService {
+//!     // ...
+//! #   fn service_id(&self) -> u16 { 1000 }
+//! #   fn service_name(&self) -> &str { "MyService" }
+//! #   fn state_hash(&self, _: &Snapshot) -> Vec<Hash> { vec![] }
+//! #   fn tx_from_raw(&self, _: RawTransaction) -> Result<Box<Transaction>, encoding::Error> {
+//! #       unimplemented!()
+//! #   }
+//!
+//!     fn public_api_handler(&self, context: &ApiContext) -> Option<Box<iron::Handler>> {
+//!         let mut api = ServiceApi::new();
+//!         api.set_transactions::<Any>();
+//!         api.insert(BALANCE_SPEC, Endpoint::new(balance));
+//!         Some(IronAdapter::new(context.clone()).create_handler(api))
+//!     }
+//! }
+//! # fn main() { }
+//! ```
+
+use serde::Serialize;
+use serde::de::DeserializeOwned;
+use serde_json::{self, Value};
+
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::{fmt, io};
+
+use blockchain::{Blockchain, SendError, Transaction};
+use crypto::{Hash, PublicKey};
+use messages::RawMessage;
+use storage::Snapshot;
+
+/// The specification for the "standard" transaction sink.
+///
+/// Transaction sinks can be added to service APIs using the [`set_transactions`]
+/// method in `ServiceApi`; see it for more details.
+///
+/// [`add_transactions`]: struct.ServiceApi.html#method.set_transactions
+pub const TRANSACTIONS: Spec = Spec { id: "transactions" };
+
+/// API-related errors.
+#[derive(Debug, Fail)]
```

Comment:
In according to failure release plan is more suitable to use `display_derive` instead of use `[fail(display = "")]`
https://boats.gitlab.io/blog/post/2018-02-22-failure-1.0/